File: examples/07-aws-containers/python/s3_reference_payload.py
"""Keep Workflow history small: pass S3 references (URIs), not file contents,
between steps. The Activities do the heavy I/O; the Workflow only sequences URIs.

The Python port of s3_reference_payload.java. boto3 is imported lazily so this
module py_compiles without the AWS SDK installed.
"""

from dataclasses import dataclass
from datetime import timedelta

from temporalio import activity, workflow


@dataclass
class TransformRequest:
    input_s3_uri: str
    output_prefix: str


@dataclass
class TransformResult:
    output_s3_uri: str
    row_count: int


def _s3_client():
    import boto3  # lazy: AWS SDK may be absent offline

    return boto3.client("s3")


@activity.defn
async def transform(input_s3_uri: str, output_prefix: str) -> str:
    # Read from input_s3_uri, write the transformed object under output_prefix,
    # and return the *new* URI - never the bytes.
    s3 = _s3_client()  # noqa: F841  (illustrative; AWS call elided)
    return f"{output_prefix}/transformed.parquet"


@activity.defn
async def count_rows(output_s3_uri: str) -> int:
    s3 = _s3_client()  # noqa: F841
    return 0  # illustrative


@workflow.defn
class TransformWorkflow:
    @workflow.run
    async def run(self, request: TransformRequest) -> TransformResult:
        # Pass S3 references, not file contents, to keep history lean.
        output_uri = await workflow.execute_activity(
            transform,
            args=[request.input_s3_uri, request.output_prefix],
            start_to_close_timeout=timedelta(minutes=5),
        )
        row_count = await workflow.execute_activity(
            count_rows,
            output_uri,
            start_to_close_timeout=timedelta(minutes=2),
        )
        return TransformResult(output_s3_uri=output_uri, row_count=row_count)

File: examples/07-aws-containers/go/s3_reference_payload.go
// Keep Workflow history small: pass S3 references (URIs), not file contents,
// between steps. The Activities do the heavy I/O; the Workflow only sequences URIs.
//
// The Go port of s3_reference_payload.java. The AWS S3 calls sit behind an S3API
// interface so the Temporal usage compiles without the AWS SDK.
package aws

import (
	"context"
	"time"

	"go.temporal.io/sdk/workflow"
)

// TransformRequest and TransformResult carry S3 URIs, never bytes.
type TransformRequest struct {
	InputS3URI   string
	OutputPrefix string
}

type TransformResult struct {
	OutputS3URI string
	RowCount    int64
}

// S3API is the slice of the AWS S3 client these Activities need.
type S3API interface {
	Copy(ctx context.Context, srcURI, dstURI string) error
	RowCount(ctx context.Context, uri string) (int64, error)
}

// TransformActivities wraps an S3API so the Activities stay testable.
type TransformActivities struct{ S3 S3API }

// Transform reads from input, writes under outputPrefix, and returns the *new*
// URI - never the bytes.
func (a *TransformActivities) Transform(ctx context.Context, inputS3URI, outputPrefix string) (string, error) {
	out := outputPrefix + "/transformed.parquet"
	if err := a.S3.Copy(ctx, inputS3URI, out); err != nil {
		return "", err
	}
	return out, nil
}

// CountRows returns the row count of an output object.
func (a *TransformActivities) CountRows(ctx context.Context, outputS3URI string) (int64, error) {
	return a.S3.RowCount(ctx, outputS3URI)
}

// TransformWorkflow sequences the two Activities, passing only S3 references so
// Workflow history stays lean.
func TransformWorkflow(ctx workflow.Context, req TransformRequest) (TransformResult, error) {
	ctx = workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
		StartToCloseTimeout: 5 * time.Minute,
	})

	var a *TransformActivities
	var outputURI string
	if err := workflow.ExecuteActivity(ctx, a.Transform, req.InputS3URI, req.OutputPrefix).
		Get(ctx, &outputURI); err != nil {
		return TransformResult{}, err
	}

	var rowCount int64
	if err := workflow.ExecuteActivity(ctx, a.CountRows, outputURI).Get(ctx, &rowCount); err != nil {
		return TransformResult{}, err
	}
	return TransformResult{OutputS3URI: outputURI, RowCount: rowCount}, nil
}
